flink-clients/src/main/java/org/apache/flink/client/program/JobClientLazy.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.client.program;
+
+import org.apache.flink.api.common.JobClient;
+import org.apache.flink.api.common.JobExecutionResult;
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.runtime.client.JobExecutionException;
+
+import java.util.Map;
+
+/**
+ * A detached job client which lazily initiates the cluster connection.
+ */
+public class JobClientLazy implements JobClient {
```

Comment:
The distinction between `JobClientEager` and `JobClientLazy` feels a little bit clumsy. Can't we get rid of one them and simply have a `JobClientImpl`? The only place where `JobClientLazy` is returned is when calling `submitJobDetached`. I think in this case, you don't expect to get a `JobClient` back because it is submitted in detached mode.